chall08/edramire.py

```diff
@@ -0,0 +1,138 @@
+#! /usr/bin/env python3
+
+import sys
+
+
+class Minefield:
+    def __init__(self):
+        self.rows = 0
+        self.cols = 0
+        self.lines = []
+        self.loaded = False
+        self.exit = None
+        self.robot = None
+        self.valid = True
+
+    def validate(self, line_str: str):
+        len_line = len(line_str)
+        if len_line < 3 or (self.rows > 0 and len_line != self.cols):
+            return False
+        if any(c not in '+*ME0' for c in line_str):
+            return False
+        e = sum((1 if c == 'E' else 0 for c in line_str), 0)
+        if e > 1 or (e == 1 and self.exit is not None):
+            return False
+        m = sum((1 if c == 'M' else 0 for c in line_str[1:len_line]), 0)
+        if m > 1 or (m == 1 and self.robot is not None):
+            return False
+        if all(c in '+E' for c in line_str):
+            if e == 1 and (line_str.startswith('E') or line_str.endswith('E')):
+                return False
+            if (self.rows > 1
+                    and all(c in '+E' for c in self.lines[self.rows - 1])):
+                return False
+        elif all(c in 'M0*' for c in line_str[1:len_line - 1]):
+            if line_str[0] not in '+E' or line_str[-1] not in '+E':
+                return False
```

Comment:
Using regular expressions here would have helped you make the code more readable I think.  
First and last line regex: `^\+[\+E]+\+$`  
Middle line regex `^[\+E][0M\*]+[\+E]$`

And then you check if the count of M and E is correct.